### DISCUSSION TITLE
API difference ui.table inbetween add_rows()+remove_rows() vs update_rows()

### DISCUSSION THREAD

**Discussion by bojh** [zauberzeug/nicegui#2633] [Ideas / Feature Requests]
### Question

I just notified that there is a difference in the API for table update methods of the table:
`*rows:Dict` -vs- `rows: List[Dict]` -- Is there reason for?
see:

```py
def remove_rows(self, *rows: Dict) -> None:
    """Remove rows from the table."""
    keys = [row[self.row_key] for row in rows]
    self.rows[:] = [row for row in self.rows if row[self.row_key] not in keys]
    self.selected[:] = [row for row in self.selected if row[self.row_key] not in keys]
    self.update()

def update_rows(self, rows: List[Dict], *, clear_selection: bool = True) -> None:
    """Update rows in the table.

    :param rows: list of rows to update
    :param clear_selection: whether to clear the selection (default: True)
    """
    self.rows[:] = rows
    if clear_selection:
        self.selected.clear()
    self.update()
```

**Comment by bojh** [comment]
Just understood, that the `update_rows()` implementantion is more a `replace_all_rows()` implementation. Means the complete table is set with the given rows List[Dict]. For a simple row update, in the first step an `remove_rows()`, `add_rows()` sequence does work properly.

**Comment by falkoschindler** [comment]
Hi @bojh,

> Is there reason for?

Not really. Originally I created `add_rows` and `remove_rows` and thought it might be handy to use `*rows` as an argument, so that you can simply pass a single row without square brackets. Later we added `update_rows`, which was a PR from the community, which expects a list of rows. I guess I thought `*rows` would be confusing with the additional keyword argument this method expects.

Ultimately I think a list of rows would have been the better choice for all three methods, because it is more straightforward and less surprising. We could try to change it in version 2.0. But ideally we should auto-detect the old usage and handle it in a backward-compatible way, to avoid regression bugs in user code that are hard to spot. Something like this could work:
```py
def add_rows(self, rows: List[Dict], *args: Any) -> None:
    """Add rows to the table.

    :param rows: list of rows to add
    :param args: additional rows to add (DEPRECATED; for backwards compatibility only)
    """
    if isinstance(rows, dict):
        rows = [rows] + args
    self.rows.extend(rows)
    self.update()
```

Let's convert this discussion into a feature request.

**Reply by bojh** [reply]
Hi @falkoschindler understood. May me a feature request would be a good idea. 
some remarks:

1.  current `add_rows(*row: Dict)` will be DEPRECATED. Two new methods can take place:
 `add_row(*row)` and `add_rows(self, rows: List[Dict], *args: Any)` 
2. ... same to `remove_rows()`
3. `update_rows(self, rows: List[Dict], *, clear_selection: bool = True)` is more interesting: 
As I understood it acts currently as **replace-all-rows-with** and a selection feature. 
So the current `update_rows()` implementation should get the name something like `replace_all_rows()` and the real `update_rows(self, rows: List[Dict], *, clear_selection: bool = True)` should take the given rows find the existing rows in the current table via row_key (like the `remove_rows()` )  and replace found one with the given content....
... forgot a new method `update_row(*row: Dict)` 

THX for your reply

**Comment by falkoschindler** [comment]
This feature request has been implemented in PR #3525 and will be released in version 2.0.0.

**Comment by bojh** [comment]
Oh fine, great THX, looking forward
